### PR TITLE
[Remote Store] Add support to restore only unassigned shards of an index

### PR DESCRIPTION
**Pull request by sachinpkale**
### Description
- Currently, [remote store restore API ](https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/remote-store/index/#restoring-from-a-backup)restores all the shards of a given index irrespective of the shard state.
- This approach is not optimized for scenarios where only few of the shards are unassigned. Restore will still download data for all the shards in order to restore them.
- We also need to close the index before calling remote store restore API. This impacts the index availability during the restore.
- Ideally, we should be able to restore only those shards that are unassigned and that too without closing the index.
- In restore API execution, while marking shards to be restored, we filter shards based on existing [ShardRoutingState](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/ShardRoutingState.java).
- We pick only shards that are in UNASSIGNED state as shards to be restored.
- User do not need to provide shard info in the API request. API signature would remain same which accepts index/indices to be restored. API implementation internally picks up only unassigned shards to restore.
- We also provided a query parameter restore_all_shards which, as the name suggests, will restore all the shards of the given index. As some of the shards can be active, to avoid an inconsistent state, when restore_all_shards=true is passed, index needs to be in closed state.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/8836

### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by sachinpkale** [comment]
Adding tests

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java [hunk 1]
```diff
@@ -212,17 +224,55 @@ private void testRestoreFlowMultipleIndices(int numberOfIterations, boolean invo
...
+                PlainActionFuture.newFuture()
+            );
         ensureGreen(indices);
         for (String index : indices) {
             assertEquals(shardCount, getNumShards(index).totalNumShards);
             verifyRestoredData(indicesStats.get(index), true, index);
         }
     }
 
+    public void testRestoreFlowAllShardsNoRedIndex() throws InterruptedException {
```
Comment (sachinpkale) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java:244]: Need to add more integ tests to cover all the flows. I will be adding them soon as part of the same PR.

File: server/src/main/java/org/opensearch/action/admin/cluster/remotestore/restore/RestoreRemoteStoreRequest.java
Comment (BhumikaSaini-Amazon) [inline server/src/main/java/org/opensearch/action/admin/cluster/remotestore/restore/RestoreRemoteStoreRequest.java:1]: Just thinking out loud, how is the following case being handled by the proposed logic:

1. Lets say the request does not pass a `restoreAllShards` param. In that case, the default value of `false` will be considered.
2. Given point# 1, the closed index check won't be enforced.
3. But, if all shards in the index are unassigned, point# 2 could be an incorrect behaviour.

In addition to the check on `restoreAllShards` being true, should we also validate if all shards are unassigned and throw the IllegalStateException if the index is not closed?

**Review by BhumikaSaini-Amazon — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/action/admin/cluster/remotestore/restore/RestoreRemoteStoreRequest.java
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/action/admin/cluster/remotestore/restore/RestoreRemoteStoreRequest.java:1]: If `restoreAllShards` is not provided, we will restore all the unassigned shards. It is possible that all the shards of the given index are unassigned. This is expected behavior.

Passing `restoreAllShards=true` can be useful where some of the shards are still active and accepting writes. If we restore without closing the index first, it is possible that we may lose some data. That is why we are restricting the index to be in closed state with  `restoreAllShards=true`

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java [hunk 2]
```diff
@@ -144,10 +138,20 @@ private void testRestoreFlow(int numberOfIterations, boolean invokeFlush, int sh
...
+        boolean restoreAllShards = randomBoolean();
+        if (restoreAllShards) {
+            assertAcked(client().admin().indices().prepareClose(INDEX_NAME));
+        }
+        client().admin()
+            .cluster()
+            .restoreRemoteStore(
+                new RestoreRemoteStoreRequest().indices(INDEX_NAME).restoreAllShards(restoreAllShards),
+                PlainActionFuture.newFuture()
+            );
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java:150]: Makes sense. Will get it abstracted out.

[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java:150]: It is covered in `testRestoreFlowAllShardsNoRedIndex`.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java [hunk 3]
```diff
@@ -212,17 +224,55 @@ private void testRestoreFlowMultipleIndices(int numberOfIterations, boolean invo
...
+
+        PlainActionFuture<RestoreRemoteStoreResponse> future = PlainActionFuture.newFuture();
+        client().admin().cluster().restoreRemoteStore(new RestoreRemoteStoreRequest().indices(INDEX_NAME).restoreAllShards(true), future);
+        try {
+            future.get();
+        } catch (ExecutionException e) {
+            // If the request goes to co-ordinator, e.getCause() can be RemoteTransportException
+            assertTrue(e.getCause() instanceof IllegalStateException || e.getCause().getCause() instanceof IllegalStateException);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java:257]: > Does this mean that while triggering a restore with restoreAllShards as true on a red index without first closing the index will throw an exception, doing the same on a yellow/green index without closing the index first will be a no-op and we expect no exception even if this would otherwise-violate what the API expects to be a valid request?

Exception will be thrown if we trigger restore for an index in any state (green, yellow, red) without closing the index first.

> Also, is this test targeted for an empty index or do we need to add a step to index some docs?

No, forgot to add docs. Will add.

> Do we need any assertions post the restore?

At least for this test, restore will not go through. Do you mean to check if the index is in the same state as before?

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java [hunk 4]
```diff
@@ -234,21 +235,34 @@ public ClusterState execute(ClusterState currentState) {
...
+                        Map<ShardId, ShardRouting> activeInitializingShards = new HashMap<>();
+                        if (request.restoreAllShards()) {
+                            if (currentIndexMetadata.getState() != IndexMetadata.State.CLOSE) {
+                                throw new IllegalStateException(
+                                    "cannot restore index ["
+                                        + index
+                                        + "] because an open index "
+                                        + "with same name already exists in the cluster. Close the existing index"
+                                );
+                            }
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/snapshots/RestoreService.java:248]: > Should we expect the same behaviour if the index is green/yellow?

Yes

> Not directly related to this PR, but we may have to assess this behaviour for restore triggered using wildcards but only some of the matched indices being closed (assuming restoreAllShards is true).

Good point. Not just wildcards, but even if we provide multiple indices, need to think on what should be the behavior. For example, at this point, we are in a loop for all the given indices. For wildcards, they should be resolved before entering the loop so we assume that we have list of indices. 
Now, `restoreAllShards` is applicable to all the indices as it is query level parameter. Does it make sense to have it at index level? But then, what does it mean to have index level parameter for wildcards? 
Also, the behavior is similar to what we have today. In current implementation, index has to be closed before restoring. It throws an exception if one of the given indices is not closed.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (BhumikaSaini-Amazon) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java:249]: Thanks Sachin

> At least for this test, restore will not go through. Do you mean to check if the index is in the same state as before?

My bad, no restore makes sense here. I do think asserting that the index is the same as before would be a good idea (can be abstracted out into a common helper maybe, which can be reused across remote store tests). It could help us catch regressions on any unwanted index modifications for cases like this one.

File: server/src/main/java/org/opensearch/action/admin/cluster/remotestore/restore/RestoreRemoteStoreRequest.java
Comment (BhumikaSaini-Amazon) [reply server/src/main/java/org/opensearch/action/admin/cluster/remotestore/restore/RestoreRemoteStoreRequest.java:1]: > Should the default be restoreAllShards  with index in a closed state as a pre-req to match that of current semantics. When we want to restore unassigned shards only, should we have a flag unassigned_only as an explicit override?

I think keeping the more optimized restore approach (i.e. `restoreAllShards` being `false`) as the default behaviour might be better, because:
1. It does not block the other shards that are active. 
2. It would likely offer a better user experience (otherwise, could it be that post GA users are almost always preferring and using the `unassigned_only` override? If yes, reversing this may not be as easy.).

But are there any cases In which this approach would be undesirable?

**Review by BhumikaSaini-Amazon — APPROVED** [review]

**Review by gbbafna — APPROVED** [review]

File: server/src/test/java/org/opensearch/action/admin/cluster/remotestore/restore/RestoreRemoteStoreRequestTests.java
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/test/java/org/opensearch/action/admin/cluster/remotestore/restore/RestoreRemoteStoreRequestTests.java:1]: It is covered in `RemoteStoreIT.testRestoreFlowAllShardsNoRedIndex`

**Comment by sachinpkale** [comment]
Integ tests are covering new changes. Bypassing the `codecov/patch` failure.
